### PR TITLE
TINKERPOP-1583: PathRetractionStrategy retracts keys that are actually needed

### PR DESCRIPTION
**Pull request by twilmes**
`PathRetractionStrategy` was not pushing `keepLabels` down into sibling traversals of children with label requirements.  Labels were also not being pushed down into `TraversalParents` earlier in the traversal.

@okram I saw you had a `PathRetractionStrategy` issue while doing some other work.  Do you have a traversal you were getting a failure on?  I can make sure this fixes it and adds a test case if so.

`./docker/build.sh -t -i` success
```
[INFO] Apache TinkerPop .................................. SUCCESS [30.949s]
[INFO] Apache TinkerPop :: Gremlin Shaded ................ SUCCESS [9.462s]
[INFO] Apache TinkerPop :: Gremlin Core .................. SUCCESS [56.670s]
[INFO] Apache TinkerPop :: Gremlin Test .................. SUCCESS [4.887s]
[INFO] Apache TinkerPop :: Gremlin Groovy ................ SUCCESS [2:09.535s]
[INFO] Apache TinkerPop :: Gremlin Groovy Test ........... SUCCESS [4.935s]
[INFO] Apache TinkerPop :: TinkerGraph Gremlin ........... SUCCESS [3:32.013s]
[INFO] Apache TinkerPop :: Gremlin Benchmark ............. SUCCESS [32:23.091s]
[INFO] Apache TinkerPop :: Gremlin Driver ................ SUCCESS [15:56.763s]
[INFO] Apache TinkerPop :: Neo4j Gremlin ................. SUCCESS [15:53.728s]
[INFO] Apache TinkerPop :: Gremlin Server ................ SUCCESS [30:15.950s]
[INFO] Apache TinkerPop :: Gremlin Python ................ SUCCESS [32:02.364s]
[INFO] Apache TinkerPop :: Hadoop Gremlin ................ SUCCESS [22:03.412s]
[INFO] Apache TinkerPop :: Spark Gremlin ................. SUCCESS [52:17.578s]
[INFO] Apache TinkerPop :: Giraph Gremlin ................ SUCCESS [2:52:20.792s]
[INFO] Apache TinkerPop :: Gremlin Console ............... SUCCESS [32:59.765s]
[INFO] Apache TinkerPop :: Gremlin Archetype ............. SUCCESS [0.085s]
[INFO] Apache TinkerPop :: Archetype - TinkerGraph ....... SUCCESS [16:05.156s]
[INFO] Apache TinkerPop :: Archetype - Server ............ SUCCESS [10.964s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

VOTE: +1

### CODE REVIEW

**Comment by okram** [comment]
So the problem I was having was related to the interplay between `PathRetractionStrategy` and `MatchPredicateStrategy`. You can't have `PathRetractionStrategy` without `MatchPredicateStrategy`.

The test that fails is:


```
g.V().match(
  as("a").in("sungBy").as("b"),
  as("a").in("writtenBy").as("c"),
  as("b").out("writtenBy").as("d"))
    .where(as("c").out("sungBy").as("d"))
    .where(as("d").has("name", "Garcia"));
```

If I take out `MatchPredicateStrategy` in the TinkerPop computer suite and only have `PathRetractionStrategy`, the test fails.

If you could figure out how to make it so this dependency doesn't exist, that would be great. Again, this is a `GraphComputer` execution (and it also happens with `GraphActors`).

**Comment by twilmes** [comment]
It looks like when I run this against tp32 without `MatchPredicateStrategy`, the `a` and `b` keys are missing from the results.  I think I figured out what the issue is.  Without `MatchPredicateStrategy`, the wheres are not integrated into the `MatchStep` so `PathRetractionStrategy` sees that `a` and `b` are not referenced after the match so they are dropped.  I'll make an update so that if this pattern is seen, the match start and end labels will be preserved in the same manner as they would be with a traversal where the `MatchStep` was the terminating step.

**Comment by twilmes** [comment]
Made a small update to remove a dependency between `PathRetractionStrategy` and `MatchPredicateStrategy`.  I should also note, this PR addresses TINKERPOP-1597.  TINKERPOP-1597 was entered against 3.3.0 and a small change to one of `PathRetractionStrategy` test cases will be required when tp32 is merged into master due to the differing implementation of the `optional`.

`docker/build.sh -t -i` ran last night with success with this last commit.

VOTE: +1

**Comment by dkuppitz** [comment]
This also fixes [TINKERPOP-1597](https://issues.apache.org/jira/browse/TINKERPOP-1597), but the test I've added can only be included in the 3.3.x line, right?

**Comment by dkuppitz** [comment]
VOTE: +1

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/PathRetractionStrategyTest.java [hunk 1]
```diff
@@ -37,14 +37,8 @@
...
 
-import static org.apache.tinkerpop.gremlin.process.traversal.P.eq;
-import static org.apache.tinkerpop.gremlin.process.traversal.P.gte;
-import static org.apache.tinkerpop.gremlin.process.traversal.P.neq;
-import static org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.__.as;
-import static org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.__.limit;
-import static org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.__.out;
-import static org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.__.select;
-import static org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.__.values;
+import static org.apache.tinkerpop.gremlin.process.traversal.P.*;
```
Comment (spmallette) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/PathRetractionStrategyTest.java]: your IDE inserted a wildcard - could you please fix the formattng @twilmes

Comment (twilmes) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/PathRetractionStrategyTest.java]: Installed a new IDE and never turned wildcard imports off.  Good catch!

**Comment by okram** [comment]
I just tested the `PathRetractionStrategy` code in the GraphActors branch and the `akka-gremlin/` test suite passes.

*** Sidenote: can you finalize your variables if possible?

VOTE +1.

**Pull request merged**